### PR TITLE
make custom partitioners work with large msg support

### PR DESCRIPTION
**Pull request by radai-rosenblatt**

### CODE REVIEW

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java [hunk 1]
```diff
@@ -70,6 +73,10 @@
 
   public static final String VALUE_SERIALIZER_CLASS_DOC = "The value serializer class";
 
+  public static final String PARTITIONER_CLASS_CONFIG_DOC = "The partitioner class";
+
+  public static final String MAX_BLOCK_MS_DOC = "see max.block.ms is kafka producer documentation";
```
Comment (jonlee2) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java]: is->in

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java [hunk 2]
```diff
@@ -100,7 +107,9 @@
         .define(MAX_MESSAGE_SEGMENT_BYTES_CONFIG, Type.INT, "800000", Importance.MEDIUM, MAX_MESSAGE_SEGMENT_BYTES_DOC)
         .define(AUDITOR_CLASS_CONFIG, Type.CLASS, NoOpAuditor.class.getName(), Importance.MEDIUM, AUDITOR_CLASS_DOC)
         .define(KEY_SERIALIZER_CLASS_CONFIG, Type.CLASS, ByteArraySerializer.class.getName(), Importance.MEDIUM, KEY_SERIALIZER_CLASS_DOC)
-        .define(VALUE_SERIALIZER_CLASS_CONFIG, Type.CLASS, ByteArraySerializer.class.getName(), Importance.MEDIUM, VALUE_SERIALIZER_CLASS_DOC)
+        .define(VALUE_SERIALIZER_CLASS_CONFIG, Type.CLASS, null, Importance.MEDIUM, VALUE_SERIALIZER_CLASS_DOC) //null by default!
```
Comment (jonlee2) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java]: Why did you change the default for value serializer?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java [hunk 3]
```diff
@@ -271,6 +292,24 @@ private LiKafkaProducerImpl(LiKafkaProducerConfig configs,
         _auditor.record(auditToken, topic, timestamp, 1L, 0L, AuditType.ATTEMPT);
         throw t;
       }
+
+      if (partition == null && _partitioner != null) {
```
Comment (jonlee2) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: what happens if partition == null and _partitioner == null?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java (hunk 2 above)
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java]: the default partitioner hashes the key, which we know works. in that case i can avoid all the reflection calls at runtime - this way users only pay the cost for this if they need it

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java (hunk 3 above)
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: that means the client (the underlying one) is using the default partitioner, which we know is safe

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java (hunk 2 above)
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerConfig.java]: unintended. thanks for catching this

File: integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java [hunk 4]
```diff
@@ -195,5 +209,77 @@ public void close() {
     assertEquals(1, messageCount);
   }
 
+  @Test
+  public void testCustomPartitioningOfLargeMessage() throws Exception {
+    Random random = new Random();
```
Comment (smccauliff) [inline integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java]: Use fixed random number seed for test reproducibility.  Or rather... why even bother with this random partitions business.

File: integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java [hunk 5]
```diff
@@ -195,5 +209,77 @@ public void close() {
...
+
+    AdminClient adminClient = createRawAdminClient(null);
+    CreateTopicsResult createTopicResult = adminClient.createTopics(Collections.singletonList(new NewTopic(topic, numPartitions, (short) 1)));
+    createTopicResult.all().get(1, TimeUnit.MINUTES);
+    TopicDescription topicDescription = adminClient.describeTopics(Collections.singletonList(topic)).all().get(1, TimeUnit.MINUTES).get(topic);
+    Assert.assertEquals(topicDescription.partitions().size(), numPartitions, "expected topic to have "
+        + numPartitions + " partitions. instead get " + topicDescription.partitions());
+
+    Properties producerPropertyOverrides = new Properties();
+    producerPropertyOverrides.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, RoundRobinPartitioner.class.getCanonicalName());
```
Comment (smccauliff) [inline integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java]: Since this partitioner starts at zero check that all the message segments end up on partition zero.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java [hunk 6]
```diff
@@ -187,17 +195,30 @@ private LiKafkaProducerImpl(LiKafkaProducerConfig configs,
...
       _keySerializer.configure(configs.originals(), true);
       // Instantiate the key serializer if necessary.
       _valueSerializer = valueSerializer != null ? valueSerializer
-          : configs.getConfiguredInstance(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+          : configs.getConfiguredInstance(LiKafkaProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
       _valueSerializer.configure(configs.originals(), false);
 
+      _partitioner = configs.getConfiguredInstance(LiKafkaProducerConfig.PARTITIONER_CLASS_CONFIG, Partitioner.class);
+      _maxBlockMs = configs.getLong(LiKafkaProducerConfig.MAX_BLOCK_MS_CONFIG);
+      if (_partitioner != null) {
```
Comment (smccauliff) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: Can javdoc be added here do explain why all the reflection is needed?

File: integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java [hunk 7]
```diff
@@ -195,5 +209,79 @@ public void close() {
...
+    TopicDescription topicDescription = adminClient.describeTopics(Collections.singletonList(topic)).all().get(1, TimeUnit.MINUTES).get(topic);
+    Assert.assertEquals(topicDescription.partitions().size(), numPartitions, "expected topic to have "
+        + numPartitions + " partitions. instead got " + topicDescription.partitions() + ". seed is " + seed);
+
+    Properties producerPropertyOverrides = new Properties();
+    producerPropertyOverrides.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, RoundRobinPartitioner.class.getCanonicalName());
+    producerPropertyOverrides.put(LiKafkaProducerConfig.LARGE_MESSAGE_ENABLED_CONFIG, "true");
+
+    int msgSize = (int) (1.5 * DEFAULT_MAX_SEGMENT_BYTES);
+    int numExpectedRecords = (int) Math.ceil(((double) msgSize) / DEFAULT_MAX_SEGMENT_BYTES);
```
Comment (viswamy) [inline integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java]: Won't this be always equal to 2?

File: integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java [hunk 8]
```diff
@@ -195,5 +209,79 @@ public void close() {
...
+        for (ConsumerRecord<byte[], byte[]> rec : records) {
+          partToCount.compute(rec.partition(), (k, v) -> {
+            if (v == null) {
+              return 1;
+            }
+            return v + 1;
+          });
+        }
+      }
+      Assert.assertEquals(recordsRead, numExpectedRecords, "only read " + recordsRead + " within timeout. seed is " + seed);
```
Comment (viswamy) [inline integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java]: This assertion can be merged with line#265. 

Assert.assertEquals(partToCount.values.iterator.next(), numExpectedRecrods, "...")

File: integration-tests/src/test/java/com/linkedin/kafka/clients/producer/RoundRobinPartitioner.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.kafka.clients.producer.Partitioner;
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.PartitionInfo;
+
+
+/**
+ * a partitioner that sends every msg to a different partition, round-robin style
+ */
+public class RoundRobinPartitioner implements Partitioner {
+  private int counter = 0;
```
Comment (viswamy) [inline integration-tests/src/test/java/com/linkedin/kafka/clients/producer/RoundRobinPartitioner.java]: Since producers are multi-thread compliant, Should this be AtomicInteger?

Comment (radai-rosenblatt) [reply integration-tests/src/test/java/com/linkedin/kafka/clients/producer/RoundRobinPartitioner.java]: in general yes, but the particular test i have is single threaded.
